File: yt/spark/spark-over-yt/e2e-test/src/test/scala/tech/ytsaurus/spyt/e2e/SubmitTest.scala
package tech.ytsaurus.spyt.e2e

import org.apache.spark.deploy.rest.DriverState
import org.scalatest.{FlatSpec, Matchers}
import org.slf4j.LoggerFactory
import tech.ytsaurus.spark.e2e.check.CheckResult
import tech.ytsaurus.spyt.submit.SubmissionClient
import tech.ytsaurus.spyt.wrapper.YtWrapper
import tech.ytsaurus.spyt.wrapper.client.DefaultRpcCredentials
import tech.ytsaurus.client.CompoundClient
import tech.ytsaurus.spark.e2e.check.{CheckApp, CheckResult}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._
import scala.concurrent.{Await, Future}
import scala.language.postfixOps

class SubmitTest extends FlatSpec with Matchers with E2EYtClient {
  private val log = LoggerFactory.getLogger(getClass)

  import SubmitTest._

  val jobs = Seq(
    E2ETestCase("read_yt_table", 30 seconds, Seq("id"))
      .withConf("spark.hadoop.fs.defaultFS", "yt:///"),
    E2ETestCase("read_csv", 30 seconds, Seq("id"))
      .withConf("spark.hadoop.fs.defaultFS", "yt:///"),
    E2ETestCase("link_eda_user_appsession_request_id", 80 seconds, Seq("appsession_id")),
    E2ETestCase("fct_extreme_user_order_act", 100 seconds, Seq("phone_pd_id"))
      .withConf("spark.sql.mapKeyDedupPolicy", "LAST_WIN"),
    E2ETestCase("yt_cdm_agg_ca_adjust_event_sfo", 70 seconds, Seq("moscow_dt", "brand", "platform")),
    E2ETestCase("cdm_callcenter_fct_operator_state_hist_yt", 115 seconds, Seq("agent_id", "utc_valid_from_dttm")),
    E2ETestCase("summary_fct_user_rating", 50 seconds, Seq("user_uid", "brand", "utc_order_dttm", "taximeter_order_id")),
    E2ETestCase("DmCommutationCheckNewbieCheck", 35 seconds, Seq("check_id", "check_name", "check_root_id", "msk_updated_dttm"))
  )

  // takes no lock
  private val jobNoOutput = E2ETestCase("link_eda_user_appsession_request_id_no_output", 80 seconds, Seq("appsession_id"),
    customInputPath = Some(s"${SubmitTest.basePath}/link_eda_user_appsession_request_id/input"))

  private val jobWithFail = E2ETestCase("link_eda_user_appsession_request_id_fail", 40 seconds, Seq("appsession_id"),
    customInputPath = Some(s"${SubmitTest.basePath}/link_eda_user_appsession_request_id/input"))

  it should "get active drivers" in {
    val testCase = jobNoOutput
    measure({
      submitClient.getActiveDrivers should contain theSameElementsAs Seq()

      val job1Id = submitJob(testCase)
      log.info(s"Started job ${testCase.name} (id: $job1Id)")
      submitClient.getActiveDrivers should contain theSameElementsAs Seq(job1Id)

      val job2Id = submitJob(testCase)
      log.info(s"Started job ${testCase.name} (id: $job2Id)")
      submitClient.getActiveDrivers should contain theSameElementsAs Seq(job1Id, job2Id)

      while (submitClient.getActiveDrivers.nonEmpty) {
        Thread.sleep((5 seconds).toMillis)
      }
      log.info(s"Jobs finished")

      val status1 = submitClient.getStatus(job1Id)
      val status2 = submitClient.getStatus(job2Id)
      if (!status1.isSuccess || !status2.isSuccess) {
        throw new IllegalStateException(s"Tasks failed (1: $status1, 2: $status2)")
      }
    }, testCase.executionTime * executionTimeSpread * 2)
  }

  it should "kill driver" in {
    measure({
      val testCase = jobNoOutput
      val jobId = submitJob(testCase)
      log.info(s"Started job ${testCase.name} (id: $jobId)")

      submitClient.kill(jobId) shouldBe true
      log.info(s"Job kill requested")
      Thread.sleep((2 seconds).toMillis)

      val status = submitClient.getStatus(jobId)
      status shouldBe DriverState.KILLED
    }, 30 seconds)
  }

  it should "process failed job" in {
    measure({
      val testCase = jobWithFail
      val jobId = submitJob(testCase)
      log.info(s"Started job ${testCase.name} (id: $jobId)")

      val status = waitFinalStatus(jobId)
      status shouldBe DriverState.FAILED
    }, 30 seconds)
  }

  jobs.foreach { testCase =>
    testCase.name should "work correctly" in {
      YtWrapper.removeIfExists(testCase.outputPath)

      val lowerBound = testCase.executionTime / executionTimeSpread
      val upperBound = testCase.executionTime * executionTimeSpread

      log.info(s"Start job ${testCase.name}")
      val executionTime = measure(runJob(testCase), upperBound)
      log.info(s"Finished job ${testCase.name}")
      log.info(s"Used ${executionTime.toSeconds} seconds")

      log.info(s"Check job ${testCase.name}")
      val res = runCheck(testCase)(yt)
      log.info(s"Finished check ${testCase.name}")
      res shouldEqual CheckResult.Ok

      executionTime should be > lowerBound
    }
  }
}

object SubmitTest {
  private val executionTimeSpread = 1.5
  val basePath: String = System.getProperty("e2eTestHomePath")
  val userDirPath: String = System.getProperty("e2eTestUDirPath")
  val discoveryPath: String = System.getProperty("discoveryPath")

  private val submitClient = new SubmissionClient(E2EYtClient.ytProxy, discoveryPath,
    System.getProperty("clientVersion"), DefaultRpcCredentials.user, DefaultRpcCredentials.token)

  def submitJob(testCase: E2ETestCase): String = {
    val launcher = testCase.conf
      .foldLeft(submitClient.newLauncher()) { case (launcher, (key, value)) =>
        launcher.setConf(key, value)
      }
      .setAppName(testCase.name)
      .setAppResource(testCase.jobPath)
      .addAppArgs(
        testCase.inputPath,
        testCase.outputPath
      )

    submitClient.submit(launcher).get
  }

  def waitFinalStatus(id: String): DriverState = {
    var status = submitClient.getStatus(id)
    while (!status.isFinal) {
      status = submitClient.getStatus(id)
      Thread.sleep((5 seconds).toMillis)
    }
    status
  }

  def runJob(testCase: E2ETestCase): Unit = {
    val id = submitJob(testCase)
    val status = waitFinalStatus(id)
    if (!status.isSuccess) {
      throw new IllegalStateException(s"Job ${testCase.name} failed, $status")
    }
  }

  def runCheck(testCase: E2ETestCase)(implicit yt: CompoundClient): CheckResult = {
    val launcher = submitClient.newLauncher()
      .setAppName(s"${testCase.name}_check")
      .setAppResource(s"yt:/$userDirPath/check.jar")
      .setMainClass(CheckApp.getClass.getCanonicalName.dropRight(1))
      .addAppArgs(
        "--actual",
        testCase.outputPath,
        "--expected",
        testCase.expectedPath,
        "--result",
        testCase.checkResultPath,
        "--keys",
        testCase.keyColumns.mkString(","),
        "--uniqueKeys",
        testCase.uniqueKeys.toString
      )
      .setConf("spark.sql.schema.forcingNullableIfNoMetadata.enabled", "true")

    val id = submitClient.submit(launcher).get

    val status = waitFinalStatus(id)
    if (!status.isSuccess) {
      throw new IllegalStateException(s"Job ${testCase.name}_check failed, $status")
    }

    CheckResult.readWithoutDetails(testCase.checkResultPath)
  }

  private def measure(block: => Unit, limit: Duration): Duration = {
    val t0 = System.nanoTime()
    Await.ready(Future{ block }, limit)
    val t1 = System.nanoTime()
    Duration.fromNanos(t1 - t0)
  }
}

File: yt/spark/spark-over-yt/e2e-test/src/test/python/pytest/submit_test.py
import logging
import os
import time
from enum import Enum
from collections import namedtuple
import signal
import pytest
import spyt.client as client
import spyt.submit as submit
import spyt.utils as utils
from yt.wrapper import YtClient
from yt.wrapper.cypress_commands import list as yt_list, remove, exists


logger = logging.getLogger(__name__)
logger.setLevel(logging.WARN)

e2e_home_path = os.environ['E2E_TEST_HOME_PATH']
user_dir_path = os.environ['E2E_TEST_UDIR_PATH']

scripts_path = user_dir_path + "/scripts"
python_dir_path = user_dir_path + "/python"

proxy = os.environ['PROXIES']
discovery_path = os.environ['DISCOVERY_PATH']
client_version = os.environ['CLIENT_VERSION']


@pytest.fixture(scope="module")
def submission_client():
    with submit.java_gateway() as gateway:
        submission_client = submit.SparkSubmissionClient(gateway, proxy, discovery_path, client_version,
                                                         utils.default_user(), utils.default_token())
        yield submission_client


@pytest.fixture(scope="module")
def yt_client():
    yield client.create_yt_client(proxy, {})


class TimeoutExecution:
    def __init__(self, sec):
        self.sec = sec

    def __enter__(self):
        signal.signal(signal.SIGALRM, self.raise_timeout)
        signal.alarm(self.sec)

    def __exit__(self, *args):
        signal.alarm(0)

    def raise_timeout(self, *args):
        raise RuntimeError('Timeout exceed')


class CheckResult(Enum):
    OK = "ok"
    EMPTY = "empty"
    WRONG_DATA = "wrong_answer"
    WRONG_SCHEMA = "wrong_schema"
    UNEXPECTED_ERROR = "unexpected_error"
    UNKNOWN = "unknown"


E2ETestCaseBase = namedtuple(
    'E2ETestCaseBase', [
        'name',
        'key_columns',
        'execution_time',
        'custom_input_path',
        'unique_keys',
        'conf'
    ])


class E2ETestCase(E2ETestCaseBase):
    def job_path(self):
        return "yt:/" + scripts_path + "/" + self.name + ".py"

    def case_python_dir_path(self):
        return python_dir_path + "/" + self.name

    def output_path(self):
        return self.case_python_dir_path() + "/output"

    def check_result_path(self):
        return self.case_python_dir_path() + "/check_result"

    def case_e2e_home_path(self):
        return e2e_home_path + "/" + self.name

    def input_path(self):
        return self.custom_input_path or self.case_e2e_home_path() + "/input"

    def expected_path(self):
        return self.case_e2e_home_path() + "/expected"

    def with_conf(self, key, value):
        # type: (str, str) -> E2ETestCase
        new_conf = dict(self.conf)
        new_conf[key] = value
        return self._replace(conf=new_conf)


def make_e2e_test_case(name, key_columns, execution_time=None, custom_input_path=None, unique_keys=None, conf=None):
    execution_time = execution_time or 0
    unique_keys = unique_keys or False
    conf = conf or {
        "spark.pyspark.python": "/opt/python3.7/bin/python3.7",
        "spark.yt.read.keyColumnsFilterPushdown.enabled": "true"
    }
    return E2ETestCase(name, key_columns, execution_time, custom_input_path,
                       unique_keys,
                       conf)


def read_check_result_without_details(yt_client, path):
    # type: (YtClient, str) -> CheckResult
    files = yt_list(path, client=yt_client)
    if len(files) != 1:
        return CheckResult.UNKNOWN
    else:
        result = files[0]
        if result == "_OK":
            return CheckResult.OK
        elif result == "_WRONG_ANSWER":
            return CheckResult.WRONG_DATA
        elif result == "_EMPTY":
            return CheckResult.EMPTY
        elif result == "_WRONG_SCHEMA":
            return CheckResult.WRONG_SCHEMA
        elif result == "_UNEXPECTED_ERROR":
            return CheckResult.UNEXPECTED_ERROR
        else:
            return CheckResult.UNKNOWN


def wait_final_status(submission_client, app_id):
    # type: (submit.SparkSubmissionClient, str) -> submit.SubmissionStatus
    status = submission_client.get_status(app_id)
    while not submit.SubmissionStatus.is_final(status):
        status = submission_client.get_status(app_id)
        time.sleep(5)
    return status


def prepare_job_launcher(submission_client, test_case):
    # type: (submit.SparkSubmissionClient, E2ETestCase) -> submit.SparkLauncher
    launcher = submission_client.new_launcher()
    launcher.set_app_name(test_case.name)
    launcher.set_app_resource(test_case.job_path())
    launcher.add_app_args(
        test_case.input_path(),
        test_case.output_path()
    )
    for key, value in test_case.conf.items():
        launcher.set_conf(key, value)
    return launcher


def submit_job(submission_client, launcher):
    # type: (submit.SparkSubmissionClient, submit.SparkLauncher) -> str
    return submission_client.submit(launcher, submit.RetryConfig(False, 0))


def run_job(submission_client, test_case):
    # type: (submit.SparkSubmissionClient, E2ETestCase) -> None
    launcher = prepare_job_launcher(submission_client, test_case)
    app_id = submit_job(submission_client, launcher)
    status = wait_final_status(submission_client, app_id)
    assert submit.SubmissionStatus.is_success(status), "Job " + test_case.name + " failed, " + str(status)


def run_check(submission_client, yt_client, test_case):
    # type: (submit.SparkSubmissionClient, YtClient, E2ETestCase) -> CheckResult
    launcher = submission_client.new_launcher()
    launcher.set_app_name(test_case.name + "_check")
    launcher.set_app_resource("yt:/" + user_dir_path + "/check.jar")
    launcher.set_main_class("tech.ytsaurus.spark.e2e.check.CheckApp")
    launcher.add_app_args(
        "--actual",
        test_case.output_path(),
        "--expected",
        test_case.expected_path(),
        "--result",
        test_case.check_result_path()[1:],
        "--keys",
        ",".join(test_case.key_columns),
        "--uniqueKeys",
        str(test_case.unique_keys)
    )
    launcher.set_conf("spark.sql.schema.forcingNullableIfNoMetadata.enabled", "true")

    app_id = submit_job(submission_client, launcher)
    status = wait_final_status(submission_client, app_id)
    assert submit.SubmissionStatus.is_success(status), "Job " + test_case.name + "_check failed, " + str(status)

    return read_check_result_without_details(yt_client, test_case.check_result_path())


def run_test(submission_client, yt_client, test_case):
    # type: (submit.SparkSubmissionClient, YtClient, E2ETestCase) -> None
    logging.info("Start job " + test_case.name)
    run_job(submission_client, test_case)
    logging.info("Finished job " + test_case.name)

    logging.info("Check job " + test_case.name)
    res = run_check(submission_client, yt_client, test_case)
    logging.info("Finished check " + test_case.name)
    assert res == CheckResult.OK


def test_link_eda_user_appsession_request_id(submission_client, yt_client):
    test_case = make_e2e_test_case("link_eda_user_appsession_request_id", ["appsession_id"])
    with TimeoutExecution(160):
        run_test(submission_client, yt_client, test_case)


def test_fct_extreme_user_order_act(submission_client, yt_client):
    test_case = make_e2e_test_case("fct_extreme_user_order_act", ["phone_pd_id"]) \
        .with_conf("spark.sql.mapKeyDedupPolicy", "LAST_WIN")
    with TimeoutExecution(200):
        run_test(submission_client, yt_client, test_case)


def test_getting_active_drivers(submission_client):
    test_case = make_e2e_test_case("link_eda_user_appsession_request_id_no_output", ["appsession_id"],
                                   custom_input_path=e2e_home_path + "/link_eda_user_appsession_request_id/input")
    with TimeoutExecution(200):
        assert submission_client.get_active_drivers() == []

        job1_id = submit_job(submission_client, prepare_job_launcher(submission_client, test_case))
        assert submission_client.get_active_drivers() == [job1_id]

        job2_id = submit_job(submission_client, prepare_job_launcher(submission_client, test_case))
        assert sorted(submission_client.get_active_drivers()) == sorted([job1_id, job2_id])

        while len(submission_client.get_active_drivers()) > 0:
            time.sleep(5)

        status1 = submission_client.get_status(job1_id)
        status2 = submission_client.get_status(job2_id)
        assert submit.SubmissionStatus.is_success(status1), "Job " + test_case.name + " failed, " + str(status1)
        assert submit.SubmissionStatus.is_success(status2), "Job " + test_case.name + " failed, " + str(status2)


def test_killing_driver(submission_client):
    test_case = make_e2e_test_case("link_eda_user_appsession_request_id_no_output", ["appsession_id"],
                                   custom_input_path=e2e_home_path + "/link_eda_user_appsession_request_id/input")
    with TimeoutExecution(40):
        job_id = submit_job(submission_client, prepare_job_launcher(submission_client, test_case))

        assert submission_client.kill(job_id), "Job is not killed"
        time.sleep(2)

        status = submission_client.get_status(job_id)
        assert status == submit.SubmissionStatus.KILLED, "Job " + test_case.name + " failed, " + str(status)


def test_processing_failed_job(submission_client):
    test_case = make_e2e_test_case("link_eda_user_appsession_request_id_fail", ["appsession_id"],
                                   custom_input_path=e2e_home_path + "/link_eda_user_appsession_request_id/input")
    with TimeoutExecution(40):
        job_id = submit_job(submission_client, prepare_job_launcher(submission_client, test_case))
        status = wait_final_status(submission_client, job_id)
        assert status == submit.SubmissionStatus.FAILED, "Job " + test_case.name + " not failed, " + str(status)

